metastore/src/java/org/apache/hadoop/hive/metastore/HMSMetricsListener.java

```diff
@@ -35,47 +37,57 @@
  */
 public class HMSMetricsListener extends MetaStoreEventListener {
 
-  public static final Logger LOGGER = LoggerFactory.getLogger(HMSMetricsListener.class);
-  private Metrics metrics;
+  private static final Logger LOGGER = LoggerFactory.getLogger(HMSMetricsListener.class);
+
+  private Counter createdDatabases, deletedDatabases, createdTables, deletedTables, createdParts,
```

Comment:
The existing metrics design replaces a well known interface (codahale) with a Hive specific one.  This isn't good.  (I know it was done for a reason, but we shouldn't continue it.)  It is better to let developers work directly with the codahale interface, particularly since that means as it grows and develops we will be able to follow it without keeping up our own interface.

The metric names are globalized in MetricConstants.java.  So the instance variables are just local references to the metric.  If class X and Y both call Metrics.getOrCreateCounter("x") they'll both get the same counter.

I did originally make Metrics a singleton class that tracks all the metrics as you suggest, but I found I was constantly accessing it as Metrics.getInstance().getOrCreateCounter("x") so it seemed easier to make getOrCreateCounter a static method itself instead.

Reply:
The advantage of using Hive specific interface is in case we decide to use a different metrics library it would be easy to plugin it in by implementing the interface. However, I am not sure how likely is that to happen but would be good to keep the design agnostic of the metrics library in use. I think it is okay if you think otherwise to keep the direct use of codahale for now.

However, I do feel that we should keep all the Counters in one Metrics implementation class in a cache like we have currently. In the current approach we have references at the instance level, so lets say if in the future we have a hundred counters for HMS, we will have a 100 variables to keep track of. I think it may be just cleaner to keep them in a cache in Metrics class and access them via the global constants defined in MetricsConstant.java. what do you think?

Reply:
I see what you're saying about tracking the gauges as local values.  It's making a mess of the code and forcing synchronization blocks in everything that has to create a gauge rather than allowing them to call getOrCreateGauge() in the way they can call getOrCreateCounter().  

What I want to preserve is an ability to access the underlying atomic without doing a lookup in a hash, for cases where the metric is in a performance sensitive area (e.g. the open session counter in HiveMetaStore).  I'd also like to be able use existing atomics in a class for the gauge (a la TxnHandler.numOpenTxns).  

I think we can achieve both of these goals.  I'll rework this to centralize gauge names but allow direct access to the underlying atomics or for the gauge creator to provide the underlying atomic.